test/cgen_test/CMakeLists.txt

```diff
@@ -1,43 +1,47 @@
-# see also http://stackoverflow.com/questions/9870162/avoid-quoting-in-cmake-add-custom-command
-separate_arguments(CUSTOM_C_FLAGS UNIX_COMMAND "${CMAKE_C_FLAGS}")
+if(CMAKE_VERSION VERSION_LESS 3.10)
+    message(WARNING "generator test disabled due to old cmake version")
+else()
+    # TODO: the use of generator expressions requires at least cmake versions 3.10 (3.5 does not, 3.10 does)
 
-add_custom_target(test_generated 
-    COMMAND ./cgen_test${CMAKE_EXECUTABLE_SUFFIX} > test_generated${CMAKE_EXECUTABLE_SUFFIX}.c
-    COMMAND ${CMAKE_C_COMPILER} ${CUSTOM_C_FLAGS} test_generated${CMAKE_EXECUTABLE_SUFFIX}.c -c
-        -I${CMAKE_SOURCE_DIR}/include WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/test/cgen_test
+    # Compilation of the generated code tests many import edge cases
+    # in the parser and code generator but due to CMake limitations,
+    # custom target dependencies only work for Make build targets.
+    #
+    # expansion of flags results in quotes the compiler won't eat,
+    # separating arguments should fix this, but not sure how portable it is.
+    # see also http://stackoverflow.com/questions/9870162/avoid-quoting-in-cmake-add-custom-command
+
+    add_custom_command(
+        OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/test_generated${CMAKE_EXECUTABLE_SUFFIX}.c"
+        COMMAND cgen_test > "${CMAKE_CURRENT_BINARY_DIR}/test_generated${CMAKE_EXECUTABLE_SUFFIX}.c"
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/test/cgen_test
+        DEPENDS cgen_test
     )
-add_dependencies(test_generated cgen_test)
-
-# Might be related to:
-# https://cmake.org/Bug/view.php?id=14963#c37230
-# https://github.com/ninja-build/ninja/issues/760
-if(${CMAKE_MAKE_PROGRAM} MATCHES make)
-# this is now also broken for make - the system include path is not
-# visible so build fails on <assert.h> not found in the custom build
-# stage where CMAKE_C_COMPILER uses a compiler call that has this
-# behavior
-#add_test(test_generated ${CMAKE_MAKE_PROGRAM} test_generated)
-endif(${CMAKE_MAKE_PROGRAM} MATCHES make)
+
+    if(NOT MSVC)
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-error=unused-const-variable")
+    endif()
+    add_library(test_generated_objects OBJECT
+        "${CMAKE_CURRENT_BINARY_DIR}/test_generated${CMAKE_EXECUTABLE_SUFFIX}.c"
+    )
+    target_include_directories(test_generated_objects PRIVATE "${PROJECT_SOURCE_DIR}/include")
+    set_target_properties(test_generated_objects PROPERTIES EXCLUDE_FROM_ALL ON)
+
+    add_custom_target(test_generated
+        COMMAND "${CMAKE_COMMAND}" -E remove -f "${CMAKE_CURRENT_BINARY_DIR}/test_generated${CMAKE_EXECUTABLE_SUFFIX}.c" $<TARGET_OBJECTS:test_generated_objects>
+        COMMAND "${CMAKE_COMMAND}" --build "${PROJECT_BINARY_DIR}" --target test_generated_objects --config $<CONFIG> --verbose
+    )
+
+    add_test(NAME test_generated COMMAND ${CMAKE_COMMAND} --build "${PROJECT_BINARY_DIR}" --target test_generated --config $<CONFIG>)
+endif()
```

Comment:
@mikkelfj 
Looks like I forgot to insert a final newline in some final lines.
Would you be interested in [adding an editorconfig file](https://editorconfig.org/) to the repository?
It's a small file auto-configuring a lot of editors with basic formatting settings.

Reply:
I use editorconfig in other projects, so fine with me. tabs = 4 spaces.

Reply:
Is it fine to use:
```
root = true

[*]
indent_style = space
indent_size = 4
```
Are there binary files committed?

Reply:
Not sure what root means.
Yes there are binary files committed. You might want to look for .gitattr or something.
Newline translation can also cause a problem in some text files due to diff in some test scripts - gitattr should fix that for windows - but this was done a long time ago.
You might want to make sure that the various shell script drivers keep working after these changes.
This includes benchmarks.

Reply:
BTW: and somewhat unrelated: I have found diff to be a fantastic test tool. Instead of creating hundreds of small test cases you  just have a larger test driver that produce output and you save and commit the output and diff against a test run. If you are friendly, you make sure to print some context in the output so the error can found when diff breaks.

Reply:
Notably the flatc compatibility test uses binary flatbuffers for reference testing.

Reply:
But to answer you question, I think the editorconfig is fine, as far as I can tell. There might be some special cases to handle as per the above.

Reply:
The editor looks in all parent directories for `.editorconfig` files.
`root` instructs the editor to not look in parent directories.
This makes sure that a `.editorconfig` in e.g. a users' directory does not cause unwanted changes.

To avoid any unintended changes, I modified the `.editorconfig` to:
```
root = true

[{*.c, *.h, *.cpp, *.bfs, *.sh, *.cmake, *.cmake.in, *.txt}]
indent_style = space
indent_size = 4
insert_final_newline = true
trim_trailing_whitespace = true

[*.yml]
indent_style = space
indent_size = 2
insert_final_newline = true
trim_trailing_whitespace = true
```
I assume the editor will only apply these rules to text files, not binary files.
But I'm not sure.

Reply:
I think `*.json` should be added to `*.yml`. Are there any yml files?
And `*.fbs` to the source file list.

Maybe you right wrt. binary files, but otherwise, from memory:
`*.bin` and `*.bfbs` are binary files, the latter being binary schema files, for former typically unspecified flatbuffer files.

Reply:
Uh, and what are `*.bfs` ? if you meant `*.bfbs` they shouln't be there, as mentioned above.

Reply:
There are none of those bin extensions and the flatc_compat dir has some strange extensions for binary and json for historical reasons, important from the google repo. Nonetheless, flatcc outputs .bfbs files with the --schema option.

Reply:
The `*.yml` files are the CI sources (appveyor + travis + github actions).
Yes, I meant to do `*.fbs`. Oops.

Reply:
Hmm, setting the <packageName>_ROOT environment or cmake variable (that we use to find the build arch flatcc compiler when cross-compiling was introduced only with CMake 3.12 :-(.

Run `cmake --help-policy CMP0074` for more information.

So we have to increase cmake_minimum_required to 3.12 or modify the cmake config files.

Reply:
And even when changing `cmake_minimum_required` to 3.12 I can't get cross-compiling to work. It looks like `find_package` is smart enough to skip packages that are not in the same architecture as the host architecture.

See following output when running cmake with --debug-find:

```
  find_package considered the following locations for the Config module:

    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/usr/flatccCliConfig.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/usr/flatcccli-config.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/flatccCliConfig.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/flatcccli-config.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/usr/flatccCliConfig.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/usr/flatcccli-config.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/flatccCliConfig.cmake
    /home/wdobbe/install/dynniq-yoctosdk/cortexa9hf/3.0.0/sysroots/cortexa9t2hf-neon-poky-linux-gnueabi/flatcccli-config.cmake
    /home/wdobbe/build/flatcc_x86_64/lib64/cmake/flatcccli/flatccCliConfig.cmake
    /home/wdobbe/build/flatcc_x86_64/lib64/cmake/flatcccli/flatcccli-config.cmake
    /home/wdobbe/build/flatcc_x86_64/flatccCliConfig.cmake
    /home/wdobbe/build/flatcc_x86_64/flatcccli-config.cmake
    /home/wdobbe/build/flatcc_cortexa9/flatccCliConfig.cmake
    /home/wdobbe/build/flatcc_cortexa9/flatcccli-config.cmake
    /home/wdobbe/build/flatcc_cortexa9/lib/cmake/flatcccli/flatccCliConfig.cmake
    /home/wdobbe/build/flatcc_cortexa9/lib/cmake/flatcccli/flatcccli-config.cmake

  The file was found at

    /home/wdobbe/build/flatcc_cortexa9/lib/cmake/flatcccli/flatcccli-config.cmake
```
find_package skips the flatccCli config file installed in /home/wdobbe/build/flatcc_x86_64/lib64/cmake/flatcccli/flatcccli-config.cmake and then continues and finally selects /home/wdobbe/build/flatcc_cortexa9/lib/cmake/flatcccli/flatcccli-config.cmake.

Reply:
Maybe go back to the very first solution where I used an environment variable in function `flatcc_generate_sources` to override the flatcc cli executable ?

Reply:
You need a higher minimum cmake version only when you're cross building.

Reply:
I'll look into cross building tomorrow.
Maybe allow an overload e.g. `FLATCC_CLI` to be used as executable.